### PR TITLE
Allow specification of default template values

### LINKED ISSUE #46: RFC: Sources of values
e.g. "looking at cli params, env vars, global config files, or, in the end, by prompting the user"

### LINKED ISSUE #389: FEA: specify `template_values_file` in multiple places
This issue takes over from an old attempt of my own (#296), and here's the idea...

What I'd like to end up with is:

1. Be able to specify default `template_values_file` in the global `.cargo-generate.toml` file
    * Maybe specify values directly in a `[values]` section?
2. Override values with `template_values_file`, definable for each defined favorite.
    * Maybe specify values directly in a `[favorites.<name>.values]` section?
3. Override with `template_values_file` from the CLI.
    * Maybe add support for `CARGO_GENERATE_<name>` style environment variables here?
    * Maybe add `--define`|`-d` as CLI argument to specify individual values?

In my mind, this should make the "value" system more complete, in that we can specify them at several well defined layers.

Before charging ahead though, I'd like to know if there's any comments/ideas/objections?

@sassman any good comments - knowing the history of #296 which basically is a failed and less well defined attempt at this?

### PR DESCRIPTION
**Pull request by taurr**
Fixes #389

Seems to also close #46

### CODE REVIEW

File: src/favorites.rs [hunk 1]
```diff
@@ -42,34 +48,48 @@ pub(crate) fn list_favorites(app_config: &AppConfig, args: &Args) -> Result<()>
...
-    let (git, branch, subfolder) = app_config
+    let (values, git, branch, subfolder) = app_config
         .favorites
-        .get(favorite_name.as_str())
+        .as_ref()
+        .and_then(|f| f.get(favorite_name.as_str()))
         .map_or_else(
             || {
-                info!(
+                warn!(
```
Comment (sassman) [inline src/favorites.rs:71]: question: I wonder why that is now a warning? I think a favourite config is still optional, hence I would stick with an info message.

File: src/git.rs [hunk 2]
```diff
@@ -231,7 +231,7 @@ fn git_clone_all(project_dir: &Path, args: GitConfig) -> Result<String> {
                 return Err(e.into());
             }
 
-            info!("Template does not seem to be a git repository, using as a plain folder");
+            warn!("Template does not seem to be a git repository, using as a plain folder");
```
Comment (sassman) [inline src/git.rs:234]: note: there the warning is reasonable and fine with me

File: tests/integration/favorites.rs [hunk 3]
```diff
@@ -7,7 +7,7 @@ use assert_cmd::prelude::*;
 use indoc::indoc;
 use std::path::PathBuf;
 
-fn create_template(name: &str) -> Project {
+fn create_template(description: &str) -> Project {
```
Comment (sassman) [inline tests/integration/favorites.rs:10]: suggestion: I found it more reasonable to name those parameters along their meaning or name within the template, here we have the `project-name` not the `description`, I suggest to stick with the naming

File: README.md [hunk 4]
```diff
@@ -335,10 +335,28 @@ cargo generate template-above
...
+Default values may be specified in the config file (specified with the `--config` flag, or in the default config file `$CARGO_HOME/cargo-generate`)
+
+**Example config file:**
+
+```toml
+[values]
+placeholder1 = "default value"
+
+[favorites.my_favorite]
+git = some template git repo
```
Comment (sassman) [inline README.md:353]: suggestion: provide a real value to be more specific, including the right type + syntax

**Review by sassman — CHANGES_REQUESTED** [review]
please find some nits inline, it would be nice if you have a look

File: src/favorites.rs (hunk 1 above)
Comment (taurr) [reply src/favorites.rs:71]: hmm, I just renamed the `info!` macro, as it really was outputting a warning, so I didn't change the meaning/output here.

I could however change it, so we just write it as info ??

File: tests/integration/favorites.rs (hunk 3 above)
Comment (taurr) [reply tests/integration/favorites.rs:10]: The argument is actually the description!

`project-name` is not being substituted here, but left for the template engine. `description` however is. Please notice this is inside a `format!` macro, thus the `{{project-name}} is ignored, while the `{}` is not.

File: src/favorites.rs (hunk 1 above)
Comment (sassman) [reply src/favorites.rs:71]: I see, have not recognized it.

File: tests/integration/favorites.rs (hunk 3 above)
Comment (sassman) [reply tests/integration/favorites.rs:10]: Thanks for clarifying

File: README.md [hunk 5]
```diff
@@ -335,10 +335,28 @@ cargo generate template-above
...
+Default values may be specified in the config file (specified with the `--config` flag, or in the default config file `$CARGO_HOME/cargo-generate`)
+
+**Example config file:**
+
+```toml
+[values]
+placeholder1 = "default value"
+
+[favorites.my_favorite]
+git = https://github.com/githubusername/mytemplate.git
```
Comment (sassman) [inline README.md:353]: I believe this needs to be put in double quotes.

**Review by sassman — APPROVED** [review]

File: README.md (hunk 5 above)
Comment (taurr) [reply README.md:353]: I believe you are right :-)

**Comment by taurr** [comment]
Fixed and rebased

**Review by sassman — APPROVED** [review]

**Pull request merged**